tachyon/math/matrix/matrix_operations.h

```diff
@@ -0,0 +1,160 @@
+math::Vector<F> MulMatVecSerial(const Eigen::MatrixBase<Derived>& matrix,
+                                const Eigen::MatrixBase<Derived2>& vector) {
+  static_assert(std::is_same_v<F, typename Derived2::Scalar>);
+
+  math::Vector<F> ret = math::Vector<F>::Constant(vector.size(), F::Zero());
+  for (Eigen::Index i = 0; i < matrix.rows(); ++i) {
+    for (Eigen::Index j = 0; j < matrix.cols(); ++j) {
+      ret[i] += matrix(i, j) * vector[j];
+    }
+  }
+  return ret;
+}
+
+template <typename Derived, typename Derived2, enum Eigen::AccessorLevels Level,
+          typename F = typename Derived::Scalar>
+math::Vector<F> MulMatVec(
+    const Eigen::MatrixBase<Derived>& matrix,
+    const Eigen::DenseCoeffsBase<Derived2, Level>& vector) {
+  static_assert(std::is_same_v<F, typename Derived2::Scalar>);
+
+  math::Vector<F> ret = math::Vector<F>::Constant(vector.size(), F::Zero());
+  if (vector.rows() == 1) {
+    OPENMP_PARALLEL_FOR(Eigen::Index i = 0; i < matrix.rows(); ++i) {
+      for (Eigen::Index j = 0; j < matrix.cols(); ++j) {
+        ret[i] += matrix(i, j) * vector(0, j);
+      }
+    }
+  } else if (vector.cols() == 1) {
+    OPENMP_PARALLEL_FOR(Eigen::Index i = 0; i < matrix.rows(); ++i) {
+      for (Eigen::Index j = 0; j < matrix.cols(); ++j) {
+        ret[i] += matrix(i, j) * vector(j, 0);
+      }
+    }
+  } else {
+    NOTREACHED();
+  }
+  return ret;
+}
+
+template <typename Derived, typename Derived2, enum Eigen::AccessorLevels Level,
+          typename F = typename Derived::Scalar>
+math::Vector<F> MulMatVecSerial(
+    const Eigen::MatrixBase<Derived>& matrix,
+    const Eigen::DenseCoeffsBase<Derived2, Level>& vector) {
+  static_assert(std::is_same_v<F, typename Derived2::Scalar>);
+
+  math::Vector<F> ret = math::Vector<F>::Constant(vector.size(), F::Zero());
+  if (vector.rows() == 1) {
+    for (Eigen::Index i = 0; i < matrix.rows(); ++i) {
+      for (Eigen::Index j = 0; j < matrix.cols(); ++j) {
+        ret[i] += matrix(i, j) * vector(0, j);
+      }
+    }
+  } else if (vector.cols() == 1) {
+    for (Eigen::Index i = 0; i < matrix.rows(); ++i) {
+      for (Eigen::Index j = 0; j < matrix.cols(); ++j) {
+        ret[i] += matrix(i, j) * vector(j, 0);
+      }
+    }
+  } else {
+    NOTREACHED();
+  }
+  return ret;
+}
+
+template <typename Derived, typename Derived2,
+          typename F = typename Derived::Scalar>
+math::Matrix<F> MulMatMat(const Eigen::MatrixBase<Derived>& matrix,
+                          const Eigen::MatrixBase<Derived2>& matrix2) {
+  static_assert(std::is_same_v<F, typename Derived2::Scalar>);
+  CHECK_EQ(matrix.cols(), matrix2.rows());
+
+  math::Matrix<F> ret =
+      math::Matrix<F>::Constant(matrix.rows(), matrix2.cols(), F::Zero());
+  OPENMP_PARALLEL_FOR(Eigen::Index i = 0; i < matrix.rows(); ++i) {
+    for (Eigen::Index j = 0; j < matrix.cols(); ++j) {
+      for (Eigen::Index k = 0; k < matrix2.cols(); ++k) {
+        ret(i, k) += matrix(i, j) * matrix2(j, k);
```

Comment:
In general, the matrix multiplication operation proceeds through a for loop as shown below. Is there a special reason for changing the index?

```suggestion
        ret(i, j) += matrix(i, k) * matrix2(k, j);
```

Reply:
Nope, there's no special reason for changing the index. Just curious, why do you say this order is general? If it doesn't impact performance, is it okay to use any order?

Reply:
In my opinion, I think that fixing `ret` and loop-rolling will improve performance. The current code has a `k` index on the left side, so the location where the value on the right side is added changes at every moment.

Reply:
There is not much of a difference and it is difficult to say that it is fast in all cases, so I resolve this reveiw.

Run on (12 X 24 MHz CPU s)
CPU Caches:
  L1 Data 64 KiB
  L1 Instruction 128 KiB
  L2 Unified 4096 KiB (x12)
Load Average: 1.45, 1.43, 1.56
-----------------------------------------------------------------------
Benchmark                             Time             CPU   Iterations
-----------------------------------------------------------------------
BM_MulMatMat2<bn254::Fr>/2          158 ns          156 ns      4532798
BM_MulMatMat2<bn254::Fr>/4         1084 ns         1073 ns       652401
BM_MulMatMat2<bn254::Fr>/8         8040 ns         7971 ns        87858
BM_MulMatMat2<bn254::Fr>/16       64137 ns        63437 ns        11024
BM_MulMatMat2<bn254::Fr>/32      517366 ns       511713 ns         1358
BM_MulMatMat2<bn254::Fr>/64     4201565 ns      4172548 ns          168
BM_MulMatMat2<bn254::Fr>/128   35062504 ns     34773000 ns           20
BM_MulMatMat2<bn254::Fr>/256  316868396 ns    313892000 ns            2
BM_MulMatMat2<bn254::Fr>/512 2744872375 ns   2728341000 ns            1
BM_MulMatMat<bn254::Fr>/2           162 ns          158 ns      4441850
BM_MulMatMat<bn254::Fr>/4          1101 ns         1089 ns       640527
BM_MulMatMat<bn254::Fr>/8          8348 ns         8271 ns        84594
BM_MulMatMat<bn254::Fr>/16        66967 ns        66285 ns        10519
BM_MulMatMat<bn254::Fr>/32       527855 ns       523056 ns         1329
BM_MulMatMat<bn254::Fr>/64      4265972 ns      4228545 ns          167
BM_MulMatMat<bn254::Fr>/128    34793796 ns     34464095 ns           21
BM_MulMatMat<bn254::Fr>/256   318716937 ns    315083000 ns            2
BM_MulMatMat<bn254::Fr>/512  2851070958 ns   2836450000 ns            1